src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,931 @@
+    pCode->EmitLDARG(ARG_CALL_TARGET);
+
+    pCode->EmitCALLI(pCode->GetSigToken(pCalliSig, cbCalliSig), 2, 0);
+
+    // callTarget = newFrameEntry.NextCall;
+    pCode->EmitLDLOC(newFrameEntryLcl);
+    pCode->EmitLDFLD(FIELD__TAIL_CALL_FRAME__NEXT_CALL);
+    pCode->EmitSTARG(ARG_CALL_TARGET);
+
+    // } while (callTarget != IntPtr.Zero);
+    pCode->EmitLDARG(ARG_CALL_TARGET);
+    pCode->EmitBRTRUE(loopStart);
+
+    // }
+    pCode->EmitLEAVE(afterTryFinally);
+    pCode->EndTryBlock();
+
+    // finally {
+    pCode->BeginFinallyBlock();
+
+    // tls->Frame = prevFrame;
+    pCode->EmitLDLOC(tlsLcl);
+    pCode->EmitLDLOC(prevFrameLcl);
+    pCode->EmitSTFLD(FIELD__TAIL_CALL_TLS__FRAME);
+
+    // }
+    pCode->EmitENDFINALLY();
+    pCode->EndFinallyBlock();
+
+    // afterTryFinally:
+    pCode->EmitLabel(afterTryFinally);
+
+    // return;
+    pCode->EmitRET();
+
+    Module* mscorlib = MscorlibBinder::GetModule();
+    MethodDesc* pDispatchTailCallsMD =
+        ILStubCache::CreateAndLinkNewILStubMethodDesc(
+            MscorlibBinder::GetModule()->GetLoaderAllocator(),
+            mscorlib->GetILStubCache()->GetOrCreateStubMethodTable(mscorlib),
+            ILSTUB_TAILCALL_DISPATCH,
+            mscorlib,
+            pSig, cbSig,
+            &emptyCtx,
+            &sl);
+
+#ifdef _DEBUG
+    StackSString ilStub;
+    sl.LogILStub(CORJIT_FLAGS(), &ilStub);
+    StackScratchBuffer ssb;
+    LOG((LF_STUBS, LL_INFO1000, "TAILCALLHELP: DispatchTailCalls IL created\n"));
+    LOG((LF_STUBS, LL_INFO1000, "%s\n\n", ilStub.GetUTF8(ssb)));
+#endif
+
+    // We might waste a MethodDesc here if we lose the race, but that is very
+    // unlikely and since this initialization only happens once not a big deal.
+    InterlockedCompareExchangeT(&s_tailCallDispatcherMD, pDispatchTailCallsMD, NULL);
+    return s_tailCallDispatcherMD;
+}
+
+void TailCallHelp::CreateTailCallHelperStubs(
+    MethodDesc* pCallerMD, MethodDesc* pCalleeMD,
+    MetaSig& callSiteSig, bool virt,
+    MethodDesc** storeArgsStub, bool* storeArgsNeedsTarget,
+    MethodDesc** callTargetStub)
+{
+    STANDARD_VM_CONTRACT;
+
+#ifdef _DEBUG
+    SigFormat incSig(callSiteSig, NULL);
+    LOG((LF_STUBS, LL_INFO1000, "TAILCALLHELP: Incoming sig %s\n", incSig.GetCString()));
+#endif
+
+    // There are two cases where we will need the tailcalling site to pass us the target:
+    // * It was a calli, for obvious reasons
+    // * The target is generic (in which case it might depend on type parameters
+    // in the caller). Since the CallTarget stub is non-generic we cannot
+    // express a call to it in IL.
+    *storeArgsNeedsTarget = pCalleeMD == NULL || callSiteSig.IsGenericMethod();
+
+    // We 'attach' the tailcall stub to the target method except for the case of
+    // calli (where the callee MD will be null). We do not reuse stubs for calli
+    // tailcalls as those are presumably very rare and not worth the trouble.
+    LoaderAllocator* loaderAlloc =
```

Comment:
A nit - can you please call it pLoaderAlloc to conform to the convention used in the runtime (or even better call it pLoaderAllocator - we use that at almost all if not all places where we use loader allocator)